src/twisted/internet/interfaces.py

```diff
@@ -1090,6 +1090,16 @@ def spawnProcess(
         arguments will be decoded up to L{unicode} using the encoding given by
         L{sys.getfilesystemencoding} (C{utf8}) and given to Windows's native "wide" APIs.
 
+        On POSIX, C{spawnProcess} will use C{os.posix_spawnp} if it is available
+        and the arguments passed to C{spawnProcess} are supported. Otherwise,
+        a combination of C{os.fork} and C{os.execvpe} is used. The following
+        arguments cause C{spawnProcess} to not use C{os.posix_spawnp}:
+
+         - C{path} being not C{None} or a path that resolves to something different
+            than the current directory.
+         - C{uid} being not C{None}.
+         - C{gid} being not C{None}.
```

Comment:
These seem more like implementation details than the public interface we want to promise (consider: spawnProcess never used to use posix_spawnp at all - and it was changed to do so without introducing a new interface).

I don't know if that means these words belong somewhere else (the interface is certainly an obvious and tempting place to put them).  There is long-form documentation about using processes with Twisted which might be a good place (but it might be less visible, if that matters).  At the very least, if this info stays here, I'd like to see a note with it saying these are implementation details and subject to change without invoking the usual backwards incompatibility process.

Reply:
I think that the narrative / non-API docs are here https://docs.twisted.org/en/stable/core/howto/process.html#running-another-process


I am +1 for moving this into the implementation API documention rather than having them in the interface.

I think one options is moving this here https://docs.twistedmatrix.com/en/stable/api/twisted.internet.process.Process.html

Reply:
@exarkun @adiroiban I agree with your comments. The current addition to documentation is badly worded and its place is in implementation docs in the current form.

However, I think that there still needs to be update to the API docs. The fact that certain uses of `spawnProcess` may invoke `os.fork` under the hood became important after `os.fork` became deprecated.

For the cases where `spawnProcess` uses `os.fork` right now, `os.fork` is the only alternative. Deprecation of `os.fork` implies that it may be removed in the future. This means that the arguments of `spawnProcess` that currently require `os.fork` are sort of softly deprecated too.

Accordingly, it makes sense to encode the fact that certain uses of `spawnProcess` are guaranteed not to invoke `os.fork` into the API contract. It probably makes sense to deprecate the arguments of `spawnProcess` that cause `os.fork` to be invoked too, but I think this is a step too far right now. Maybe `os.fork` will be deprecated but not removed for a decade and in the meantime `posix_spawnp` will get the missing functionality on the platforms we care about.

Reply:
I've updated the wording to be more generic. It only specifies that under certain conditions `os.fork` *may* be used. If these conditions do not hold, then `os.fork` *will not* be used. I think this is generic enough and does not constrain the implementation in any way as at this point increasing the usage of `os.fork` should be considered a bug.